common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3774,6 +3775,485 @@ static jint NativeCrypto_EVP_AEAD_CTX_open_buf(JNIEnv* env, jclass, jlong evpAea
+                                                          jobject hpkeKeyCtxRef, jlong evpKdfRef, jlong evpAeadRef,
+                                                          jbyteArray encArray, jbyteArray infoArray) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    EVP_HPKE_CTX* ctx = fromContextObject<EVP_HPKE_CTX>(env, recipientHpkeCtxRef);
+    EVP_HPKE_KEY* key = fromContextObject<EVP_HPKE_KEY>(env, hpkeKeyCtxRef);
+
+    if (ctx == nullptr || key == nullptr) {
+        // NullPointerException thrown while calling fromContextObject
+        return;
+    }
+
+    const EVP_HPKE_KDF* kdf = reinterpret_cast<const EVP_HPKE_KDF*>(evpKdfRef);
+    const EVP_HPKE_AEAD* aead = reinterpret_cast<const EVP_HPKE_AEAD*>(evpAeadRef);
+    if (kdf == nullptr || aead == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "kdf or aead == null");
+        return;
+    }
+
+    if (encArray == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "encArray == null");
+        return;
+    }
+
+    JNI_TRACE("EVP_HPKE_CTX_setup_recipient(%p, %p, %p, %p, %p, %p)",
+               ctx, key, kdf, aead, encArray, infoArray);
+    jbyte* enc = env->GetByteArrayElements(encArray, 0);
+    size_t encLen = env->GetArrayLength(encArray);
+
+    jbyte* infoArrayElement = nullptr;
+    const uint8_t* info = nullptr;
+    size_t infoLen = 0;
+    if (infoArray != nullptr) {
+        infoArrayElement = env->GetByteArrayElements(infoArray, 0);
+        info = reinterpret_cast<const uint8_t*>(infoArrayElement);
+        infoLen = env->GetArrayLength(infoArray);
+    }
+
+    if (!EVP_HPKE_CTX_setup_recipient(/* ctx= */ ctx,
+                                      /* key= */ key,
+                                      /* kdf= */ kdf,
+                                      /* aead= */ aead,
+                                      /* enc= */ reinterpret_cast<const uint8_t*>(enc),
+                                      /* enc_len= */ encLen,
+                                      /* info= */ info,
+                                      /* info_len= */ infoLen)) {
+        env->ReleaseByteArrayElements(encArray, enc, JNI_ABORT);
+        if (infoArrayElement != nullptr) {
+            env->ReleaseByteArrayElements(infoArray, infoArrayElement, JNI_ABORT);
+        }
+        conscrypt::jniutil::throwExceptionFromBoringSSLError(env, "EVP_HPKE_CTX_setup_recipient");
+        return;
+    }
+
+    env->ReleaseByteArrayElements(encArray, enc, JNI_ABORT);
+    if (infoArrayElement != nullptr) {
+        env->ReleaseByteArrayElements(infoArray, infoArrayElement, JNI_ABORT);
+    }
+}
+
+// Defining EVP_HPKE_KEM struct with only the field needed to call the
+// function "EVP_HPKE_CTX_setup_sender_with_seed_for_testing" using "kem->seed_len"
+struct evp_hpke_kem_st {
+  size_t seed_len;
+};
+
+static jbyteArray NativeCrypto_EVP_HPKE_CTX_setup_sender_with_seed(JNIEnv* env, jclass,
```

Comment:
This is a testing-only interface and needs to retain the `for_testing` suffix to capture this.

Edit: I see you're using this outside testing. This is not supported and needs to be redesigned to not need this. This PR *cannot* land until this is fixed.